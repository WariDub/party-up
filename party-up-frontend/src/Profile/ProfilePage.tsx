import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import NavBar from '../Search/components/NavBar';
import EditProfileForm from './components/EditProfileForm';
import FriendList from './components/FriendList';
import User from './models/user.model';
import * as jwtDecode from 'jwt-decode';
import { JwtPayload } from '../Auth/interfaces/jwt-payload.interface';
import { UsersService } from './users.service';

export interface ProfilePageProps extends RouteComponentProps {}

export interface ProfilePageState {
  user: User | null;
  friends: User[];
}

const ProfilePage = class extends React.Component<ProfilePageProps, ProfilePageState> {
  usersService = new UsersService();

  constructor(props: ProfilePageProps) {
    super(props);

    this.state = {
      user: null,
      friends: [],
    };
  }

  async componentDidMount() {
    await this.fetchAndSetUser();
    const { user } = this.state;
    if (user) {
      await this.fetchAndSetFriends(user);
    }
  }

  render(): JSX.Element {
    const { history, location, match } = this.props;
    const { user, friends } = this.state;

    return (
      <>
        <NavBar
          history={history}
          location={location}
          match={match}
          showSearchBar={false}
          onQueryResult={null}
        />
        {user ? (
          <EditProfileForm history={history} location={location} match={match} user={user} />
        ) : null}
        <FriendList user={user} friends={friends} />
      </>
    );
  }

  fetchAndSetUser = async (): Promise<void> => {
    const accessToken = localStorage.getItem('token');
    if (!accessToken) {
      return console.error('missing access token');
    }

    const payload: JwtPayload = jwtDecode.default(accessToken);

    try {
      const user = await this.usersService.getUser(payload.uid);
      if (user) {
        this.setState({ user });
      }
    } catch (e) {
      console.error(e);
    }
  };

  fetchAndSetFriends = async (user: User): Promise<void> => {
    const operations = [];

    for (const friend of user.friends) {
      operations.push(this.usersService.getUser(friend));
    }

    try {
      const results = await Promise.all(operations);
      const friends: User[] = [];
      for (const result of results) {
        if (result === null) {
          continue;
        }
        friends.push(result);
      }

      this.setState({ friends });
    } catch (e) {
      console.error(e);
    }
  };
};

export default ProfilePage;
