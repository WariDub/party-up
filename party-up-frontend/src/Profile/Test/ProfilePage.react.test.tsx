import  ProfilePage  from '../ProfilePage';
import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { shallow } from 'enzyme';
import { createBrowserHistory } from 'history';
import { matchPath } from 'react-router';
import renderer from 'react-test-renderer';

describe('ProfilePage', () => {

    test('It should mount', () => {

        let history = createBrowserHistory();

        history.push('/');

        let location = history.location;

        let match = matchPath('/', {
            path:'/',
            exact: true,
            strict: false
        });

        let component = shallow(<ProfilePage history = { history } location = { location } match = { match }/>);

        expect(component.length).toBe(1);
    });

    test('It should mount', () => {

        let history = createBrowserHistory();

        history.push('/');

        let location = history.location;

        let match = matchPath('/', {
            path:'/',
            exact: true,
            strict: false
        });

        let component = renderer.create(<ProfilePage history = { history } location = { location } match = { match }/>).toJSON();

        expect(component).toMatchSnapshot();
    });
});