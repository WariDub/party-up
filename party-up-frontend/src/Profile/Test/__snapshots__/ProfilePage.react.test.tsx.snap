// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage It should mount 1`] = `
Array [
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorDefault MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h6
        className="MuiTypography-root MuiTypography-h6"
      >
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
          href="/"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          PartyUp
        </a>
      </h6>
      <div
        className="MuiBox-root MuiBox-root-1"
      />
      <div
        className="MuiGrid-root MuiGrid-justify-xs-flex-end"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Profile
          </span>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-justify-xs-flex-end"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Logout
          </span>
        </button>
      </div>
    </div>
  </header>,
  <div
    className="MuiBox-root MuiBox-root-2"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
      >
        <ul
          className="MuiGridList-root"
          style={
            Object {
              "margin": -4,
            }
          }
        />
      </div>
    </div>
  </div>,
]
`;
